drivers/ieee802154/ieee802154_cc13xx_cc26xx.c

```diff
@@ -0,0 +1,735 @@
+	if (sys_read32(CCFG_BASE + CCFG_O_IEEE_MAC_0) != 0xFFFFFFFF &&
+	    sys_read32(CCFG_BASE + CCFG_O_IEEE_MAC_1) != 0xFFFFFFFF) {
+		mac = (u8_t *)(CCFG_BASE + CCFG_O_IEEE_MAC_0);
+	}
+	memcpy(&drv_data->ext_addr, mac, sizeof(drv_data->ext_addr));
+	drv_data->short_addr = 0x0000;
+	drv_data->pan_id = 0x0000;
+
+	/* Setup circular RX queue (TRM 25.3.2.7) */
+	memset(&drv_data->rx_entry[0], 0, sizeof(drv_data->rx_entry[0]));
+	memset(&drv_data->rx_entry[1], 0, sizeof(drv_data->rx_entry[1]));
+
+	drv_data->rx_entry[0].pNextEntry = (u8_t *)&drv_data->rx_entry[1];
+	drv_data->rx_entry[0].config.type = DATA_ENTRY_TYPE_PTR;
+	drv_data->rx_entry[0].config.lenSz = 1;
+	drv_data->rx_entry[0].length = sizeof(drv_data->rx_data[0]);
+	drv_data->rx_entry[0].pData = drv_data->rx_data[0];
+
+	drv_data->rx_entry[1].pNextEntry = (u8_t *)&drv_data->rx_entry[0];
+	drv_data->rx_entry[1].config.type = DATA_ENTRY_TYPE_PTR;
+	drv_data->rx_entry[1].config.lenSz = 1;
+	drv_data->rx_entry[1].length = sizeof(drv_data->rx_data[1]);
+	drv_data->rx_entry[1].pData = drv_data->rx_data[1];
+
+	drv_data->rx_queue.pCurrEntry = (u8_t *)&drv_data->rx_entry[0];
+	drv_data->rx_queue.pLastEntry = NULL;
+
+	k_sem_init(&drv_data->tx_done, 0, 1);
+	k_sem_init(&drv_data->rx_done, 0, 1);
+	k_sem_init(&drv_data->ack, 0, 1);
+
+	k_thread_create(&drv_data->rx_thread, drv_data->rx_stack,
+			K_THREAD_STACK_SIZEOF(drv_data->rx_stack),
+			ieee802154_cc13xx_cc26xx_rx, dev, NULL, NULL,
+			K_PRIO_COOP(2), 0, K_NO_WAIT);
+}
+
+static int ieee802154_cc13xx_cc26xx_init(struct device *dev)
+{
+	struct ieee802154_cc13xx_cc26xx_data *drv_data = get_dev_data(dev);
+
+	/* Apply RF patches */
+#ifdef CONFIG_SOC_SERIES_CC13X2_CC26X2
+	rf_patch_cpe_ieee_802_15_4();
+#endif /* CONFIG_SOC_SERIES_CC13X2_CC26X2 */
+
+	/* Set high frequency clock source (TRM 7.5.1.1) */
+	bool set_osc_hf = OSCClockSourceGet(OSC_SRC_CLK_HF) != OSC_XOSC_HF;
+	if (set_osc_hf) {
+		/* Enable 48 MHz crystal oscillator */
+		OSCClockSourceSet(OSC_SRC_CLK_HF, OSC_XOSC_HF);
+	}
+
+	ieee802154_cc13xx_cc26xx_data_init(dev);
+
+	/* Switch high frequency clock source (TRM 7.5.1.1) */
+	if (set_osc_hf) {
+		/* Wait for oscillator to be stable */
+		while (!OSCHfSourceReady()) {
+			continue;
+		}
+		/* Switch high frequency clock source to crystal oscillator */
+		u32_t key = irq_lock();
+		OSCHfSourceSwitch();
+		irq_unlock(key);
+	}
+
+	/* Enable power domain and wait to avoid faults */
+	PRCMPowerDomainOn(PRCM_DOMAIN_RFCORE);
+	while (PRCMPowerDomainStatus(PRCM_DOMAIN_RFCORE) ==
+	       PRCM_DOMAIN_POWER_OFF) {
+		continue;
+	}
+
+	/* Enable clock domain and wait for PRCM registers to update */
+	PRCMDomainEnable(PRCM_DOMAIN_RFCORE);
+	PRCMLoadSet();
+	while (!PRCMLoadGet()) {
+		continue;
+	}
+
+	__ASSERT_NO_MSG(PRCMRfReady());
+
+	/* Enable CPE0 and CPE1 interrupts */
+	RFCCpeIntDisable(0xFFFFFFFF);
+
+	IRQ_CONNECT(CC13XX_CC26XX_CPE0_IRQ, 0,
+		    ieee802154_cc13xx_cc26xx_cpe0_isr,
+		    DEVICE_GET(ieee802154_cc13xx_cc26xx), 0);
```

Comment:
you already have the device given as a parameter to this function, so no need of DEVICE_GET (thus no need of DEVICE_DECLARE)

Reply:
The arguments to IRQ_CONNECT must be known at compile time, so I get an error to that effect if I try to use the device pointer parameter. I could also define the driver data above the init function and use `&ieee802154_cc13xx_cc26xx_data` instead, but I thought this was typically done using DEVICE_DECLARE and DEVICE_GET.

Reply:
True. It's been a while since I did a driver. I used to have a dedicated config structure which would hold an instance specific irq config function which in turn would use DEVICE_GET only, since it was positioned below the DEVICE_AND_API_INIT() macro. That was necessary since the driver could have 2+ instance. Which is not necessary for your driver I guess (it's not going to happen anytime soon to have 2+ instance of this driver).

Nevermind then.